### DISCUSSION TITLE
How can I use fast API to process this req-reply pattern ?

### DISCUSSION THREAD

**Discussion by yichong96** [fastapi/fastapi#4271] [Questions] [answered]
I would like to implement this request-reply pattern on Fast API. This is an image search service using fast API. I would like to know what is the best way to do so ?

1. Client sends a HTTP post request containing a query image to fast API server. Fast API server sends back an "OK received" response. Fast API server will then start finding the best image results matching the query.

2. Once the top-K results are found, fast API will then call K HTTP post request back to the client that originally sent the query image. Each post request will contain one image result. 

I read that Fast API allows for processing Background Task. However, I wouldn't consider this as a background task ?

Thank you for the help in advance !

**Comment by jgould22** [comment, marked as answer]
Image searching is a relatively CPU bound task and would not be well suited for running in async python. 

As a result you should probably split this into two parts. The http api and a separate service to do your image searching.

You can create your task with fast api and feed your image search service with something like Celery. 

The FastAPI portion will create the task and put it on the celery queue and your image search service can consume from the queue and emit your webhooks when it is has completed its search.

I would also recommend you don't emit a post for each result. If you have a lot of results it will be a lot of requests and that can cause problems. You should probably just return a list of results in a single request with an ID or link for the image. The client can then collect the images as required from an endpoint in your http interface.

**Comment by yichong96** [comment]
@jgould22 Thanks for the reply and suggestions. Yup, I am sending the images for processing to a separate service. The web server can continue to receive other HTTP requests while awaiting for the image search task to complete. For the image search task, I am sending the images to Triton Inference server. I have not heard about Celery. 

> You can create your task with fast api and feed your image search service with something like Celery. The FastAPI portion will create the task and put it on the celery queue and your image search service can consume from the queue and emit your webhooks when it is has completed its search.

I will research on this and see if it is possible. Thank you for the guidance !
